Three component phase separation for ultramassive white dwarfs, part 2

evbauer opened the pull request:
Updated code in phase_separation.f90 and a new wd_o_ne_3_phase test suite.

mcastrotapia committed:
Three component phase separation for ultramassive white dwarfs (#860)

Updated code in phase_separation.f90 and a new wd_o_ne_3_phase test suite.

Commit:
fix fortitude issues

Commit:
fix script permissions, more fortitude fixes

mcastrotapia commented:
Thanks for the fixes and for reviewing the PR! I will merge the branch into my fork, and I'll be attentive to comments/questions and updates.

pmocz committed:
[ci skip] fix license statements

evbauer commented on star/private/phase_separation.f90:
```diff
@@ -113,137 +114,216 @@ subroutine do_2component_phase_separation(s, dt, components, ierr)
                if(s% phase(k) <= eos_phase_boundary) then
-                  s% crystal_core_boundary_mass = s% m(k+1)
-                  exit
+                  if (s% crystal_core_boundary_mass>s% m(k+1)) then
```
What is this meant to catch? It seems there must be some subtle bug here I was missing.

mcastrotapia replied:
I think I was trying to prevent the core from moving inwards in case the extra heating suddenly changes the energy in both phases and the phase parameter. However, I think it's not necessary since the lines where I added the following:

`if (s% phase(s% nz) < eos_phase_boundary) then !!! prevent to move the core size inwards if the core is suddently "melted" leaving everything liquid under phi<0.9`

            `if (s% crystal_core_boundary_mass>0d0)then
               s% crystal_core_boundary_mass=s% crystal_core_boundary_mass
               return
            else 
               s% crystal_core_boundary_mass = 0d0
               return
            end if
         end if`

 must be enough for that

evbauer commented on star/private/phase_separation.f90:
```diff
@@ -40,41 +39,48 @@ subroutine do_phase_separation(s, dt, ierr)
          end if
       end subroutine do_phase_separation
 
       subroutine do_2component_phase_separation(s, dt, components, ierr)
```
If we keep the current code structure, we'll need to rename this subroutine to something other than "do_2component...", since it is now explicitly being used to do some 3 component cases as well. I had another branch a while ago that was attempting a different branch for 3-component distillation, but that is now somewhat stale and never got to a point of being ready to merge in. I'll need to reconcile the naming scheme somehow...

mcastrotapia replied:
Okay, I agree, if you want me to help with that, please let me know :)

evbauer commented on star/private/phase_separation.f90:
```diff
@@ -40,41 +39,48 @@ subroutine do_phase_separation(s, dt, ierr)
+         if(s% phase(s% nz) < eos_phase_boundary) then !!! prevent to move the core size inwards if the core is suddently "melted" leaving everything liquid under phi<0.9
+            if (s% crystal_core_boundary_mass>0d0)then
+               s% crystal_core_boundary_mass=s% crystal_core_boundary_mass
+               return
```
I see why you might need something like this, but I'm not sure this is the best approach in general. One alternative I've experimented with is just redistributing the phase separation heating over a significant fraction of the interior. That way you avoid transient local heating spikes that might melt the core, but still get the net right amount of energy deposited in the interior, which seems to work for capturing the effect on cooling rate. @mcastrotapia do you think that could be a viable alternative here?

mcastrotapia replied:
I think it is a good alternative. At some point, I tried something similar, but I didn't have success, and I ended up doing this because it seemed easier and faster to implement.

evbauer replied:
Ok, I introduced the heat redistribution in https://github.com/MESAHub/mesa/pull/863/commits/5519a13a3f8ac3a89fac035ce66d67ec138f9a7f. Let's see if that works instead.

evbauer commented on star/private/phase_separation.f90:
```diff
@@ -83,18 +89,14 @@ subroutine do_2component_phase_separation(s, dt, components, ierr)
          XC = s% xa(net_ic12,k_bound)
          XO = s% xa(net_io16,k_bound)
-         XNe = s% xa(net_ine20,k_bound)
-         ! Check that we're still in C/O or O/Ne dominated material as appropriate,
-         ! otherwise skip phase separation
-         if(components == 'CO'.and. XO + XC < 0.9d0) return
-         if(components == 'ONe'.and. XNe + XO < 0.8d0) return  ! O/Ne mixtures tend to have more byproducts of burning mixed in
```
I believe we need to bring these lines back. Is there an argument for deleting them?

They shouldn't affect the new 3-componenent functionality that we're trying to add here, and I'd like to keep the behavior of the 2-component options unchanged unless there is a compelling reason to change them.

mcastrotapia replied:
I deleted them when trying to let the code decide for itself if doing C/O or O/Ne in the case of 2-component phase separation, this is because in ultramassive WD models the core is usually O/Ne, but towards the outer layers there is a C/O dominated zone before reaching the H/He atmosphere (for example, Figure 2 of [Camisassa et al., 2019](https://www.aanda.org/articles/aa/pdf/2019/05/aa33822-18.pdf)).  In the 3-component, I was also letting the code check which were the 3 most abundant elements in the liquid zone to decide which phase diagram must be used.

evbauer replied:
The bigger concern here is when you get all the way out to the interface with the He envelope, at which point I want to make sure phase separation ends. We can return to make a final decision on this once I refactor into separate 2-component and 3-component logic.

evbauer commented on star/private/phase_separation.f90:
```diff
@@ -359,42 +715,35 @@ subroutine update_model_ (s, kc_t, kc_b, do_brunt)
         ! and related quantities on faces kc_t:kc_b)
         call set_micro_vars(s, kc_t, kc_b, &
-             skip_eos=.TRUE., skip_net=.TRUE., skip_neu=.TRUE., skip_kap=.FALSE., ierr=ierr)
+             skip_eos=.TRUE., skip_net=.TRUE., skip_neu=.TRUE., skip_kap=.TRUE., ierr=ierr)
```
I think we need to revert this change. This is now skipping the opacity update that is the point of this call. In fact, as written, I think this call to `set_micro_vars` skips everything and does nothing at all.

mcastrotapia replied:
I agree, I was mostly having convergence problems with this when using a small timestep and (I think) with the heating in the solid phase for such timesteps. That's why it would be good to implement the heating term as you suggested in one of the previous comments

evbauer commented on star/private/phase_separation.f90:
```diff
@@ -252,103 +332,379 @@ subroutine mix_outward(s,kbot,min_mix_zones)
+   call interp_evbicub_db( &
+            x1_, x2_, x1_l, num_x1, x2_l, num_x2, &
+            ilinx, iliny, deltax1_sob_f1, num_x1, ict, fval, ier)
+    dx1_=fval(1)  ! delta_x1 from 2d interpolation
```
Indentation of this block should be fixed. (Just a minor aesthetic note.)

evbauer replied:
There's another block later on with the same indentation problem as well.

mcastrotapia replied:
Okay :)

evbauer commented on star/test_suite/wd_o_ne_3_phase/src/run_star_extras.f90:
Note to self for later: update run_star_extras, rn script, etc to make this a fully-fledged test case and integrate it into do1_test_source for the test suite.

evbauer commented on star/private/phase_separation.f90:
```diff
@@ -113,137 +114,216 @@ subroutine do_2component_phase_separation(s, dt, components, ierr)
          ierr = 0
       end subroutine do_2component_phase_separation
 
       subroutine move_one_zone(s,k,components)
```
The `move_one_zone` subroutine has developed a lot of different branches that I think make it rather confusing. I can see why it developed into what it is, but I'd probably like to refactor this into two separate routines: `move_one_zone_for_2comp` that basically just retains the old functionality (plus the nice addition of accounting for Ne22), and `move_one_zone_for_3comp` that implements all the new 3-component cases.

mcastrotapia replied:
That sounds good!

evbauer commented:
We'll also need to add some documentation of the new options in controls.defaults.

Commit:
Merge branch 'main' into three_component_PS

Commit:
revert some minor changes, address some compiler warnings

Commit:
fix indentation

Commit:
fix whitespace, remove some unused variables flagged by compiler warnings

Commit:
introduce option to redistribute phase separation energy over inner half of WD

Commit:
try removing extra safeguard that PS heat smoothing should prevent needing

Commit:
simplify some diffs and remove extra crystal_core_boundary_mass check

Commit:
refactor phase separation into separate 2component and 3component flows

Commit:
rename phase separation outer loop routine

Commit:
remove some unused variables

Commit:
more unused variables

Commit:
remove some extraneous diffs

Commit:
indentation and spacing (no code changes)

Commit:
minimize diffs

Commit:
partial fixup for 3 component test case

VincentVanlaer commented on star/private/phase_separation.f90:
```diff
@@ -322,14 +457,298 @@ real(dp) function blouin_delta_xne(Xin)
+           open(unit=iounit, file='ONeNa_deltaO.dat', action='read',status='old')
+        else if  (components=='COMg') then
+           open(unit=iounit, file='COMg_deltaMg.dat', action='read',status='old')
+        end if
```
Use newunit:


```suggestion
        ! setup interpolation table for tau sob eta
        if (components=='CONe') then
           open(newunit=iounit, file='CONe_deltaO.dat', action='read',status='old')
        else if  (components=='NeOMg') then
           open(newunit=iounit, file='NeOMg_deltaO.dat', action='read',status='old')
        else if  (components=='ONeNa') then
           open(newunit=iounit, file='ONeNa_deltaO.dat', action='read',status='old')
        else if  (components=='COMg') then
           open(newunit=iounit, file='COMg_deltaMg.dat', action='read',status='old')
        end if
```

evbauer replied:
Happily, converting to hdf5 made this even cleaner. +1 for @rhdtownsend's forum HDF5 interface!

VincentVanlaer commented on star/private/phase_separation.f90:
```diff
@@ -322,14 +457,298 @@ real(dp) function blouin_delta_xne(Xin)
+           open(unit=iounit, file='ONeNa_deltaNa.dat', action='read',status='old')
+        else if  (components=='COMg') then
+           open(unit=iounit, file='COMg_deltaC.dat', action='read',status='old')
+        end if
```
Use newunit:


```suggestion
        ! setup interpolation table for x1 x2 dx1
        if (components=='CONe') then
           open(newunit=iounit, file='CONe_deltaC.dat', action='read',status='old')
        else if  (components=='NeOMg') then
           open(newunit=iounit, file='NeOMg_deltaMg.dat', action='read',status='old')
        else if  (components=='ONeNa') then
           open(newunit=iounit, file='ONeNa_deltaNa.dat', action='read',status='old')
        else if  (components=='COMg') then
           open(newunit=iounit, file='COMg_deltaC.dat', action='read',status='old')
        end if
```

Commit:
convert 3-phase PS data files to hdf5

Commit:
delete large 3-component PS .dat files

Commit:
fix index-ordering issue in h5 files for fortran-style reads

Commit:
fix memory leak when reading 3-component PS files repeatedly

Commit:
[ci skip] move phase separation data files into star_data

Commit:
Merge branch 'main' into three_component_PS

Commit:
[ci skip] setup for test-suite integration for wd_o_ne_3_phase (work in progress)

Commit:
clean up inlist_wd_o_ne_3_phase and test

Commit:
[ci optional] clean up starting model generation for wd_o_ne_3_phase, full test

Commit:
[ci skip] pgstar cleanup

Commit:
add missing test suite calls to run_star_extras

Commit:
[ci optional] document 3-component option in controls.defaults, full test

Commit:
Make WD phase separation smooth heating by default.

The rate of phase separation heating is somewhat mesh and timestep
dependent. It should get the integrated amount of energy right, but
we are not resolving the spatial and time scales over which it occurs.
This led to an undesirable behavior at small time steps where
crystallizing a zone could cause enough mixing to generate a transient
heating spike. When localized near the crystallization front, this
could re-melt previously crystallized material in an unphysical
way. The smoothing approach here redistributes this energy throughout
the inner half of the star, which should prevent re-melting, so we at
least get the right amount of energy into the stellar interior, while
we might not be precisely locating it in the right place.
This changes the default behavior somewhat, so I've included an
inlist flag in case somebody wants to turn this off to reproduce
previously published behavior.

Commit:
add missing variable declaration

evbauer commented on star/test_suite/wd_o_ne_3_phase/README.rst:
```diff
@@ -0,0 +1,23 @@
+
+The README file should end with a line that describes when the most
+recent significant changes to this test case occurred.  This gives
+both users and developers a sense of how "fresh" the information is.
```
@mcastrotapia could you fill this in or send me anything you'd like to be included here?

mcastrotapia committed:
Update README.rst (#1037)

Commit:
tweak README and include images

Commit:
Merge branch 'main' into three_component_PS [ci optional]

VincentVanlaer commented on star/private/phase_separation.f90:
```diff
@@ -40,26 +41,35 @@ subroutine do_phase_separation(s, dt, ierr)
             write(*,*) 'invalid phase_separation_option'
             stop
          end if
+
```
You will need a check for ierr here, otherwise `smooth_eps_phase_sep` will hide the error. Since neither functions actually set ierr to anything else than zero, the ierr parameter could also be removed.

VincentVanlaer commented on star/private/phase_separation.f90:
```diff
@@ -322,14 +464,351 @@ real(dp) function blouin_delta_xne(Xin)
+        phase_sep_data_dir = trim(mesa_data_dir) // '/star_data/phase_separation'
+
+        ! setup interpolation table for tau sob eta
+        if (components=='CONe') then
+           filename = trim(phase_sep_data_dir) // '/CONe_deltaO.h5'
+        else if  (components=='NeOMg') then
+           filename = trim(phase_sep_data_dir) // '/NeOMg_deltaO.h5'
+        else if  (components=='ONeNa') then
+           filename = trim(phase_sep_data_dir) // '/ONeNa_deltaO.h5'
+        else if  (components=='COMg') then
+           filename = trim(phase_sep_data_dir) // '/COMg_deltaMg.h5'
+        end if
```
I don't know how often these data reads are called. If it is on the order of 1 to 10s of times per run its probably fine. Otherwise we might want to keep this around in memory. This isn't a blocker for me to merge in since it is a new feature anyway (and disabled by default?) and leave it for a rainy day.

evbauer replied:
Good point! In principle you're definitely right that it would be better to keep these around in memory. I think they get called 100s of times per run. In practice reading these small files doesn't seem to be much of a bottleneck right now, especially after converting them to hdf5, so I think we'll leave it for a rainy day for now.

VincentVanlaer commented on star/defaults/controls.defaults:
```diff
@@ -7563,6 +7566,17 @@
+
+      ! if true, redistribute all heating associated with phase
+      ! separation evenly throughout the inner half of the white dwarf
+      ! (prevents unphysical localized spikes at small time steps)
```
Would be good to document the choice for the inner half (could be arbitrary, but I think it would nice for the user to know why).

VincentVanlaer reviewed (approved):
Left some small comments, which can be resolved easily or can be postponed. I think you should add a changelog entry, this feature certainly deserves it!

Commit:
[ci optional] have wd model generation stop slightly earlier in wd_o_ne_3_phase

Commit:
[ci optional] make wd_o_ne_3_phase more robust to atmosphere/EOS problems

Commit:
document heat redistribution for phase separation better, add check for ierr

Commit:
Merge branch 'main' into three_component_PS [ci optional]

Commit:
add changelog entry

evbauer committed:
[ci optional] fix atmosphere boundary table transition problem

evbauer committed:
Merge branch 'main' into three_component_PS

evbauer commented:
Thanks for the suggestions @VincentVanlaer! I've addressed them (other than the loading into memory suggestion, which I'd like to leave for later), and I added a changelog entry. I think this should be ready to merge at last.

I've had a bit of trouble getting the new test case to be robust to small changes in starting model generation when running all inlists, which is why there are several [ci optional] commits here while I was working that out. I've added a few things that will hopefully help it be more robust, but I'll keep an eye on the testing for a while too.
